### PR TITLE
Issue 963

### PR DESCRIPTION
**Pull request by Yuriyzabegaev**
## Proposed changes

This addresses #963 .

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [ ] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [ ] The documentation is up-to-date.
- [ ] Static typing is included in the update.
- [ ] This PR does not duplicate existing functionality.
- [ ] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: tests/geometry/test_intersections.py [hunk 1]
```diff
@@ -40,11 +41,286 @@ def test_one_and_two():
...
+    x_max = np.array([1, 3, 3])
+
+    pairs = intersections._identify_overlapping_intervals(x_min, x_max)
+
+    assert pairs.size == 4
+    assert check_pairs_contain(pairs, [0, 1])
+    assert check_pairs_contain(pairs, [1, 2])
+
+
+# ---------- Testing _intersect_pairs ----------
```
Comment (keileg) [inline tests/geometry/test_intersections.py:181]: It should be possible to parametrize this part as well.

File: tests/geometry/test_intersections.py [hunk 2]
```diff
@@ -40,11 +41,286 @@ def test_one_and_two():
...
+    combined_pairs = np.sort(
+        intersections._identify_overlapping_rectangles(x_min, x_max, y_min, y_max),
+        axis=0,
+    )
+    assert combined_pairs.shape[1] == 4
+
+    assert np.allclose(pairs_1, combined_pairs)
+
+
+# ---------- Testing split_intersecting_segments_2d ----------
```
Comment (keileg) [inline tests/geometry/test_intersections.py:267]: And here..

File: tests/utils/test_default_domains.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+"""Tests for utility functions for generation of default domains."""
+import porepy as pp
+import pytest
+
+
+class TestDomain:
```
Comment (keileg) [inline tests/utils/test_default_domains.py:6]: Please parametrize

File: tests/fracs/test_plane_fracture.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+
+@pytest.fixture
+def frac():
+    return plane_fracture.PlaneFracture(
+        np.array([[0, 1, 1, 0], [0, 0, 1, 1], [0, 0, 0, 0]]), check_convexity=False
+    )
+
+
+def test_not_vertex(frac):
```
Comment (keileg) [inline tests/fracs/test_plane_fracture.py:98]: Please parameterize these tests

File: tests/fracs/test_plane_fracture.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+def make_plane_fracture(index: int | None):
+    array = np.array([[0, 1, 0], [0, 0, 0], [0, 0, 1]])
+    return plane_fracture.PlaneFracture(array, index=index, check_convexity=False)
+
+
+def test_equal_index():
+    f1 = make_plane_fracture(index=1)
+    f2 = make_plane_fracture(index=1)
+    assert f1 == f2
```
Comment (keileg) [inline tests/fracs/test_plane_fracture.py:72]: Parametrization please

File: tests/fracs/test_plane_fracture.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
+import numpy as np
+import pytest
+
+from porepy.fracs import plane_fracture
+
+# Test computation of fracture centroids.
+
+
+def test_frac_3d_1():
```
Comment (keileg) [inline tests/fracs/test_plane_fracture.py:9]: Parametrization of this set of tests, please.

File: tests/fracs/test_fracture_utils.py [hunk 7]
```diff
@@ -31,47 +27,47 @@ def nrm(x, y):
     return is_true
 
 
-class TestFractureLength(unittest.TestCase):
+class TestFractureLength:
```
Comment (keileg) [inline tests/fracs/test_fracture_utils.py:30]: Why is this a class? I don't see the purpose.

File: tests/fracs/test_fracture_network_3d.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+    network.impose_external_boundary(domain)
+    p_known = np.array(
+        [[0.0, 0.5, 0.5, 0], [5.0 / 6, 5.0 / 6, 1, 1], [0.0, 0.0, 0.25, 0.25]]
+    )
+    assert len(network.fractures) == (1 + 6)
+    p_comp = network.fractures[0].pts
+    assert compare_arrays(p_comp, p_known, sort=True, tol=1e-5)
+
+
+class TestFractureNetwork3dBoundingBox:
```
Comment (keileg) [inline tests/fracs/test_fracture_network_3d.py:100]: Does this need to be a class?

File: tests/fracs/test_fracture_network_3d.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+    def test_single_fracture(self):
```
Comment (keileg) [inline tests/fracs/test_fracture_network_3d.py:101]: This and the next two (possibly three) methods could be parametrized, I believe?

File: tests/fracs/test_fracture_network_3d.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+
+@pytest.fixture
+def fracture() -> PlaneFracture:
+    return PlaneFracture(
+        np.array([[0, 1, 1, 0], [0.5, 0.5, 0.5, 0.5], [0, 0, 1, 1]]),
+        check_convexity=False,
+    )
+
+
+def test_completely_outside_lower(domain: Domain, fracture: PlaneFracture):
```
Comment (keileg) [inline tests/fracs/test_fracture_network_3d.py:35]: We can parametrize most of these functions, can we not?

File: tests/fracs/test_fracture_network_2d.py [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+        network_1._edges = np.vstack(
+            (network_1._edges, tag1 * np.ones(2, dtype=np.int8))
+        )
+        together = network_1.add_network(network_2)
+        known_tags = np.array([tag1, tag1, tag2, tag2])
+        assert np.all(together._edges[2] == known_tags)
+
+    # Below are tests with polytopal domains
+
+    def test_triangle_domain_no_fracs(self):
```
Comment (keileg) [inline tests/fracs/test_fracture_network_2d.py:246]: The following three tests could be parametrized - correct?

File: tests/fracs/test_fracture_network_2d.py [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+    def test_copy(self):
+        network_1 = pp.create_fracture_network(self.fracs)
+
+        copy = network_1.copy()
+        num_p = self.pts.shape[1]
+
+        network_1._pts = np.random.rand(2, num_p)
+        assert np.allclose(copy._pts, self.pts)
+
+    def test_no_snapping(self):
```
Comment (keileg) [inline tests/fracs/test_fracture_network_2d.py:298]: Can we parametrize snapping methods? I think so, but I'm not sure.

File: tests/fracs/test_fracture_network_2d.py [hunk 13]
```diff
@@ -0,0 +1,337 @@
+import numpy as np
+import porepy as pp
+from porepy.applications.test_utils.arrays import compare_arrays
+from porepy.fracs.utils import pts_edges_to_linefractures
+
+
+class TestFractureNetwork2d:
```
Comment (keileg) [inline tests/fracs/test_fracture_network_2d.py:7]: Any reason these tests should live in a class?

File: tests/fracs/test_fracture_importer.py [hunk 14]
```diff
@@ -0,0 +1,317 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
```
Comment (keileg) [inline tests/fracs/test_fracture_importer.py:3]: Please ensure the module docstring gives an overview of what is being tested here. This applies to all files.

**Review by keileg — COMMENTED** [review]
I believe this set of files will be much easier to work with with some more adaptations. In particular, there are many classes that I don't see the need for (I may be wrong, if so let me know), and there should be room for much more, and meaningful, test parametrizations.

File: tests/geometry/test_intersections.py [hunk 15]
```diff
@@ -40,11 +41,286 @@ def test_one_and_two():
...
-    triangulation = pp.intersections.triangulations(p2, p1, t2, t1)
+    triangulation = intersections.triangulations(p2, p1, t2, t1)
     assert len(triangulation) == 2
     assert triangulation[0][0] == 0
     assert triangulation[0][1] == 0
     assert triangulation[0][2] == 0.25
     assert triangulation[1][1] == 1
     assert triangulation[1][0] == 0
     assert triangulation[1][2] == 0.25
+
```
[the remark this answers is not in the document]
Comment (Yuriyzabegaev) [reply tests/geometry/test_intersections.py:52]: Agreed, doing it.

File: tests/geometry/test_intersections.py (hunk 1 above)
Comment (Yuriyzabegaev) [reply tests/geometry/test_intersections.py:181]: Agreed, doing it.

File: tests/geometry/test_intersections.py (hunk 2 above)
Comment (Yuriyzabegaev) [reply tests/geometry/test_intersections.py:267]: Doing it

File: tests/utils/test_default_domains.py (hunk 3 above)
Comment (Yuriyzabegaev) [reply tests/utils/test_default_domains.py:6]: Agreed, doing it

File: tests/fracs/test_plane_fracture.py (hunk 4 above)
Comment (Yuriyzabegaev) [reply tests/fracs/test_plane_fracture.py:98]: Agreed, doing it

File: tests/fracs/test_plane_fracture.py (hunk 5 above)
Comment (Yuriyzabegaev) [reply tests/fracs/test_plane_fracture.py:72]: Agreed, doing it

File: tests/fracs/test_plane_fracture.py (hunk 6 above)
Comment (Yuriyzabegaev) [reply tests/fracs/test_plane_fracture.py:9]: Agreed, doing it

File: tests/fracs/test_fracture_utils.py (hunk 7 above)
Comment (Yuriyzabegaev) [reply tests/fracs/test_fracture_utils.py:30]: Agreed, changing it

File: tests/fracs/test_fracture_network_3d.py (hunk 8 above)
Comment (Yuriyzabegaev) [reply tests/fracs/test_fracture_network_3d.py:100]: Agreed, changing it

File: tests/fracs/test_fracture_network_3d.py (hunk 9 above)
Comment (Yuriyzabegaev) [reply tests/fracs/test_fracture_network_3d.py:101]: Agreed, changing it

File: tests/fracs/test_fracture_network_3d.py (hunk 10 above)
Comment (Yuriyzabegaev) [reply tests/fracs/test_fracture_network_3d.py:35]: Agreed, changing it

File: tests/fracs/test_fracture_network_2d.py (hunk 11 above)
Comment (Yuriyzabegaev) [reply tests/fracs/test_fracture_network_2d.py:246]: Changing it

File: tests/fracs/test_fracture_network_2d.py (hunk 12 above)
Comment (Yuriyzabegaev) [reply tests/fracs/test_fracture_network_2d.py]: Changing it

**Review by keileg — APPROVED** [review]

**Pull request merged**